How to get signatures to use `#[implement]` for COM interfaces?

I have a file like the following:

```rs
use crate::bindings::*;
use crate::bindings::{
    Windows::Win32::Foundation::{BOOL, BSTR, POINT, RECT},
    Windows::Win32::UI::TextServices::{ITfMenu, TfLBIClick, TF_LANGBARITEMINFO},
    Windows::Win32::UI::WindowsAndMessaging::HICON,
};
use windows::{implement, Guid, IUnknown, IntoParam, Result};

#[implement(
    Windows::Win32::UI::TextServices::ITfLangBarItemButton,
    Windows::Win32::UI::TextServices::ITfSource
)]
pub struct LanguageBar {}

impl LanguageBar {
    // ITfLangBarItem

    pub unsafe fn GetInfo(&self) -> Result<TF_LANGBARITEMINFO> {}
    pub unsafe fn GetStatus(&self) -> Result<u32> {}
    pub unsafe fn Show<'a>(&self, fshow: impl IntoParam<'a, BOOL>) -> Result<()> {}
    pub unsafe fn GetTooltipString(&self) -> Result<BSTR> {}

    // ITfLangBarItemButton

    pub unsafe fn OnClick<'a>(
        &self,
        click: TfLBIClick,
        pt: impl IntoParam<'a, POINT>,
        prcarea: *const RECT,
    ) -> Result<()> {
    }
    pub unsafe fn InitMenu<'a>(&self, pmenu: impl IntoParam<'a, ITfMenu>) -> Result<()> {}
    pub unsafe fn OnMenuSelect(&self, wid: u32) -> Result<()> {}
    pub unsafe fn GetIcon(&self) -> Result<HICON> {}
    pub unsafe fn GetText(&self) -> Result<BSTR> {}

    // ITfSource

    pub unsafe fn AdviseSink(
        &self,
        riid: *const Guid,
        punk: Option<IUnknown>,
        pdwcookie: *mut u32,
    ) -> Result<()> {
    }
    pub unsafe fn UnadviseSink(&self, dwcookie: u32) -> Result<()> {}
}
```

Basically I searched the signature for each function from https://microsoft.github.io/windows-docs-rs/ and copy-pasted it. That causes quite a lot errors (except return type errors which are expected):

<details>

```
error[E0061]: this function takes 12 arguments but 11 arguments were supplied
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |  ^
   | |  |
   | |__supplied 11 arguments
   |    expected 12 arguments
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0061]: this function takes 5 arguments but 8 arguments were supplied
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |  ^
   | |  |
   | |__supplied 8 arguments
   |    expected 5 arguments
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `POINT: RuntimeType` is not satisfied
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |__^ the trait `RuntimeType` is not implemented for `POINT`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `RECT: RuntimeType` is not satisfied
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |__^ the trait `RuntimeType` is not implemented for `RECT`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0606]: casting `&*mut RECT` as `*const RECT` is invalid
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |__^
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `ITfMenu: RuntimeType` is not satisfied
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |__^ the trait `RuntimeType` is not implemented for `ITfMenu`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0061]: this function takes 0 arguments but 1 argument was supplied
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |  ^
   | |  |
   | |__supplied 1 argument
   |    expected 0 arguments
   |
note: associated function defined here
  --> ime\src\languagebar.rs:34:19
   |
34 |     pub unsafe fn GetIcon(&self) -> Result<HICON> {}
   |                   ^^^^^^^ -----
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0061]: this function takes 0 arguments but 1 argument was supplied
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |  ^
   | |  |
   | |__supplied 1 argument
   |    expected 0 arguments
   |
note: associated function defined here
  --> ime\src\languagebar.rs:35:19
   |
35 |     pub unsafe fn GetText(&self) -> Result<BSTR> {}
   |                   ^^^^^^^ -----
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0606]: casting `&*mut Guid` as `*const Guid` is invalid
  --> ime\src\languagebar.rs:9:1
   |
9  | / #[implement(
10 | |     Windows::Win32::UI::TextServices::ITfLangBarItemButton,
11 | |     Windows::Win32::UI::TextServices::ITfSource
12 | | )]
   | |__^
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

</details>

It seems such copy-pasting is not the way to go, if then what's the right way to get the signatures?

--- a/crates/gen/src/parser/method_signature.rs
+++ b/crates/gen/src/parser/method_signature.rs
@@ -480,7 +480,7 @@ impl MethodParam {
         let kind = self.signature.kind.gen_name(gen);
 
         if self.param.is_input() {
-            if self.signature.kind.is_primitive() {
+            if self.signature.kind.is_blittable() {
                 quote! { #name }
             } else {
                 quote! { &*(&#name as *const <#kind as ::windows::Abi>::Abi as *const <#kind as ::windows::Abi>::DefaultType) }

--- a/tests/implement/tests/com_chain.rs
+++ b/tests/implement/tests/com_chain.rs
@@ -21,8 +21,7 @@ impl Test {
         Ok(())
     }
 
-    // TODO: should be able to declare BOOL without a reference.
-    fn Save(&self, _: &Option<IStream>, _: &BOOL) -> Result<()> {
+    fn Save(&self, _: &Option<IStream>, _: BOOL) -> Result<()> {
         Ok(())
     }
 

